tiledb/sm/filesystem/s3.cc

```diff
@@ -947,35 +976,50 @@ Status S3::init_client() const {
   }
 #endif
 
+  if (!aws_access_key_id.empty() && !aws_secret_access_key.empty() &&
+      !aws_role_arn.empty()) {
+    return Status::S3Error(
+        "Ambiguous authentication credentials; both permanent and temporary "
+        "authentication credentials are configured");
+  }
+
   // If the user set config variables for AWS keys use them.
   if (!aws_access_key_id.empty() && !aws_secret_access_key.empty()) {
     Aws::String access_key_id(aws_access_key_id.c_str());
     Aws::String secret_access_key(aws_secret_access_key.c_str());
-    client_creds_ = std::unique_ptr<Aws::Auth::AWSCredentials>(
-        new Aws::Auth::AWSCredentials(access_key_id, secret_access_key));
-
-    // If the user has set a session token (for AWS Security Token Service)
-    // then use it:
-    //     - https://docs.aws.amazon.com/STS/latest/APIReference/Welcome.html
-    // For testing run: `aws sts get-session-token --duration-seconds 900`. See:
-    //     -
-    //     https://docs.aws.amazon.com/cli/latest/reference/sts/get-session-token.html
-    if (!aws_session_token.empty()) {
-      Aws::String session_token(aws_session_token.c_str());
-      client_creds_->SetSessionToken(session_token);
-    }
+    Aws::String session_token(
+        !aws_session_token.empty() ? aws_session_token.c_str() : "");
+    credentials_provider_ =
+        std::make_shared<Aws::Auth::SimpleAWSCredentialsProvider>(
+            access_key_id, secret_access_key, session_token);
+  } else if (!aws_role_arn.empty()) {
+    // If AWS Role ARN provided instead of access_key and secret_key,
+    // temporary credentials will be fetched by assuming this role.
+
+    Aws::String role_arn(aws_role_arn.c_str());
+    Aws::String external_id(
+        !aws_external_id.empty() ? aws_external_id.c_str() : "");
+    int load_frequency(
+        !aws_load_frequency.empty() ? std::stoi(aws_load_frequency) : 999);
```

Comment:
minor: to avoid magic numbers we can have 999 defined as a constant to better describe what it stands for.

Reply:
It is arbitrary time constant. So how about just add a comment there?

Reply:
Good point: it looks like we could use the existing constant in the SDK: `Aws::Auth:: DEFAULT_CREDS_LOAD_FREQ_SECONDS`:
https://sdk.amazonaws.com/cpp/api/LATEST/_s_t_s_assume_role_credentials_provider_8h_source.html

Reply:
Will change it to the constant